// redoc/lib/components/SecuritySchemes/SecuritySchemes.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var common_elements_1 = require("../../common-elements");
var styled_components_1 = require("../../styled-components");
var Markdown_1 = require("../Markdown/Markdown");
var AUTH_TYPES = {
    oauth2: 'OAuth2',
    apiKey: 'API Key',
    http: 'HTTP',
    openIdConnect: 'Open ID Connect',
};
var AuthTable = styled_components_1.default.table(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  ul > li {\n    margin: 0.5em 0 !important;\n  }\n\n  th {\n    text-transform: capitalize;\n  }\n"], ["\n  ul > li {\n    margin: 0.5em 0 !important;\n  }\n\n  th {\n    text-transform: capitalize;\n  }\n"])));
var OAuthFlow = /** @class */ (function (_super) {
    __extends(OAuthFlow, _super);
    function OAuthFlow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OAuthFlow.prototype.render = function () {
        var _a = this.props, type = _a.type, flow = _a.flow;
        return (React.createElement("tr", null,
            React.createElement("th", null,
                " ",
                type,
                " OAuth Flow "),
            React.createElement("td", null,
                type === 'implicit' || type === 'authorizationCode' ? (React.createElement("div", null,
                    React.createElement("strong", null, " Authorization URL: "),
                    flow.authorizationUrl)) : null,
                type === 'password' || type === 'clientCredentials' || type === 'authorizationCode' ? (React.createElement("div", null,
                    React.createElement("strong", null, " Token URL: "),
                    flow.tokenUrl)) : null,
                flow.refreshUrl && (React.createElement("div", null,
                    React.createElement("strong", null, " Refresh URL: "),
                    flow.refreshUrl)),
                React.createElement("div", null,
                    React.createElement("strong", null, " Scopes: ")),
                React.createElement("ul", null, Object.keys(flow.scopes).map(function (scope) { return (React.createElement("li", { key: scope },
                    React.createElement("code", null, scope),
                    " - ",
                    flow.scopes[scope])); })))));
    };
    return OAuthFlow;
}(React.PureComponent));
exports.OAuthFlow = OAuthFlow;
var SecurityDefs = /** @class */ (function (_super) {
    __extends(SecurityDefs, _super);
    function SecurityDefs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SecurityDefs.prototype.render = function () {
        if (!this.props.securitySchemes) {
            return null;
        }
        return (React.createElement("div", null, this.props.securitySchemes.schemes.map(function (scheme) { return (React.createElement("div", { "data-section-id": scheme.sectionId, key: scheme.id },
            React.createElement(common_elements_1.H2, null,
                React.createElement(common_elements_1.ShareLink, { href: '#' + scheme.sectionId }),
                scheme.id),
            React.createElement(Markdown_1.Markdown, { source: scheme.description || '' }),
            React.createElement(AuthTable, { className: "security-details" },
                React.createElement("tbody", null,
                    React.createElement("tr", null,
                        React.createElement("th", null, " Security scheme type: "),
                        React.createElement("td", null,
                            " ",
                            AUTH_TYPES[scheme.type] || scheme.type,
                            " ")),
                    scheme.apiKey ? (React.createElement("tr", null,
                        React.createElement("th", null,
                            " ",
                            scheme.apiKey.in,
                            " parameter name:"),
                        React.createElement("td", null,
                            " ",
                            scheme.apiKey.name,
                            " "))) : scheme.http ? ([
                        React.createElement("tr", { key: "scheme" },
                            React.createElement("th", null, " HTTP Authorization Scheme "),
                            React.createElement("td", null,
                                " ",
                                scheme.http.scheme,
                                " ")),
                        scheme.http.scheme === 'bearer' &&
                            scheme.http.bearerFormat && (React.createElement("tr", { key: "bearer" },
                            React.createElement("th", null, " Bearer format "),
                            React.createElement("td", null,
                                " \"",
                                scheme.http.bearerFormat,
                                "\" "))),
                    ]) : scheme.openId ? (React.createElement("tr", null,
                        React.createElement("th", null, " Connect URL "),
                        React.createElement("td", null,
                            React.createElement("a", { target: "_blank", href: scheme.openId.connectUrl }, scheme.openId.connectUrl)))) : scheme.flows ? (Object.keys(scheme.flows).map(function (type) { return (React.createElement(OAuthFlow, { key: type, type: type, flow: scheme.flows[type] })); })) : null)))); })));
    };
    return SecurityDefs;
}(React.PureComponent));
exports.SecurityDefs = SecurityDefs;
var templateObject_1;
