// redoc/lib/services/models/SecurityRequirement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var openapi_1 = require("../../utils/openapi");
var SecurityRequirementModel = /** @class */ (function () {
    function SecurityRequirementModel(requirement, parser) {
        var schemes = (parser.spec.components && parser.spec.components.securitySchemes) || {};
        this.schemes = Object.keys(requirement || {})
            .map(function (id) {
            var scheme = parser.deref(schemes[id]);
            var scopes = requirement[id] || [];
            if (!scheme) {
                console.warn("Non existing security scheme referenced: " + id + ". Skipping");
                return undefined;
            }
            return {
                id: id,
                sectionId: openapi_1.SECURITY_SCHEMES_SECTION + id,
                type: scheme.type,
                scopes: scopes,
            };
        })
            .filter(function (scheme) { return scheme !== undefined; });
    }
    return SecurityRequirementModel;
}());
exports.SecurityRequirementModel = SecurityRequirementModel;
