// redoc/lib/src/components/ContentItems/ContentItems.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var mobx_react_1 = require("mobx-react");
var React = require("react");
var MenuStore_1 = require("../../services/MenuStore");
var Markdown_1 = require("../Markdown/Markdown");
var common_elements_1 = require("../../common-elements");
var Operation_1 = require("../Operation/Operation");
var ContentItems = /** @class */ (function (_super) {
    __extends(ContentItems, _super);
    function ContentItems() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ContentItems.prototype.render = function () {
        var items = this.props.items;
        if (items.length === 0) {
            return null;
        }
        return items.map(function (item) { return React.createElement(ContentItem, { item: item, key: item.id }); });
    };
    ContentItems = __decorate([
        mobx_react_1.observer
    ], ContentItems);
    return ContentItems;
}(React.Component));
exports.ContentItems = ContentItems;
var ContentItem = /** @class */ (function (_super) {
    __extends(ContentItem, _super);
    function ContentItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ContentItem.prototype.render = function () {
        var item = this.props.item;
        var content;
        var type = item.type;
        switch (type) {
            case 'group':
                content = null;
                break;
            case 'tag':
                content = React.createElement(TagItem, { item: item });
                break;
            case 'section':
                return null;
            case 'operation':
                content = React.createElement(OperationItem, { item: item });
                break;
            default:
                throw new Error('Unknown item type');
        }
        return (React.createElement(React.Fragment, null,
            React.createElement("div", __assign({}, (_a = {}, _a[MenuStore_1.SECTION_ATTR] = item.id, _a)), content),
            item.items && React.createElement(ContentItems, { items: item.items })));
        var _a;
    };
    ContentItem = __decorate([
        mobx_react_1.observer
    ], ContentItem);
    return ContentItem;
}(React.Component));
exports.ContentItem = ContentItem;
var TagItem = /** @class */ (function (_super) {
    __extends(TagItem, _super);
    function TagItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TagItem.prototype.render = function () {
        var _a = this.props.item, name = _a.name, description = _a.description;
        return (React.createElement(common_elements_1.Row, null,
            React.createElement(common_elements_1.MiddlePanel, { key: "middle" },
                React.createElement(common_elements_1.H1, null,
                    React.createElement(common_elements_1.ShareLink, { href: '#' + this.props.item.getHash() }),
                    name),
                description !== undefined && React.createElement(Markdown_1.Markdown, { source: description })),
            React.createElement(common_elements_1.EmptyDarkRightPanel, { key: "right" })));
    };
    TagItem = __decorate([
        mobx_react_1.observer
    ], TagItem);
    return TagItem;
}(React.Component));
exports.TagItem = TagItem;
var OperationItem = /** @class */ (function (_super) {
    __extends(OperationItem, _super);
    function OperationItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    OperationItem.prototype.render = function () {
        return React.createElement(Operation_1.Operation, { operation: this.props.item });
    };
    OperationItem = __decorate([
        mobx_react_1.observer
    ], OperationItem);
    return OperationItem;
}(React.Component));
exports.OperationItem = OperationItem;
