// redoc/lib/src/common-elements/CodeFlask.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var styled_components_1 = require("styled-components");
var react_form_1 = require("react-form");
var CodeFlask = require("codeflask");
var codeflask_css_1 = require("codeflask/src/codeflask.css");
var forms_1 = require("./forms");
styled_components_1.injectGlobal(templateObject_1 || (templateObject_1 = __makeTemplateObject(["", ""], ["", ""])), codeflask_css_1.default);
var CodeFlaskWrapper = styled_components_1.default.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  position: relative;\n  min-height: 400px;\n  background: #ffffff;\n  border-radius: 4px;\n\n  .CodeFlask__pre {\n    box-shadow: none;\n  }\n  .CodeFlask__code {\n    color: #333;\n  }\n  .CodeFlask__code {\n    .token.attr-name,\n    .token.builtin,\n    .token.char,\n    .token.inserted,\n    .token.selector,\n    .token.string {\n      color: #018e10;\n    }\n  }\n"], ["\n  position: relative;\n  min-height: 400px;\n  background: #ffffff;\n  border-radius: 4px;\n\n  .CodeFlask__pre {\n    box-shadow: none;\n  }\n  .CodeFlask__code {\n    color: #333;\n  }\n  .CodeFlask__code {\n    .token.attr-name,\n    .token.builtin,\n    .token.char,\n    .token.inserted,\n    .token.selector,\n    .token.string {\n      color: #018e10;\n    }\n  }\n"])));
var CodeFlaskEditor = /** @class */ (function (_super) {
    __extends(CodeFlaskEditor, _super);
    function CodeFlaskEditor(props) {
        var _this = _super.call(this, props) || this;
        _this.instance = new CodeFlask('  ');
        return _this;
    }
    CodeFlaskEditor.prototype.componentDidMount = function () {
        var _this = this;
        this.instance.scaffold(this.el, false, { language: 'javascript' });
        this.instance.onUpdate(function (value) { return _this.updateValue(value); });
    };
    CodeFlaskEditor.prototype.componentWillReceiveProps = function (nextProps) {
        if (nextProps.fieldApi.getValue() !== this.instance.textarea.value) {
            this.instance.update(nextProps.fieldApi.getValue() || '');
        }
    };
    CodeFlaskEditor.prototype.updateValue = function (value) {
        this.props.fieldApi.setValue(value);
    };
    CodeFlaskEditor.prototype.render = function () {
        var _this = this;
        var fieldApi = this.props.fieldApi;
        var getError = fieldApi.getError, getTouched = fieldApi.getTouched;
        var error = getError();
        var touched = getTouched();
        return (React.createElement(React.Fragment, null,
            React.createElement(CodeFlaskWrapper, { innerRef: function (el) { return (_this.el = el); } }),
            error && touched ? React.createElement(forms_1.FieldError, null,
                " ",
                error) : null));
    };
    return CodeFlaskEditor;
}(React.Component));
exports.CodeFlaskEdit = react_form_1.FormField(CodeFlaskEditor);
var templateObject_1, templateObject_2;
