// redoc/lib/src/common-elements/perfect-scrollbar.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var perfect_scrollbar_1 = require("perfect-scrollbar");
var perfect_scrollbar_css_1 = require("perfect-scrollbar/css/perfect-scrollbar.css");
var styled_components_1 = require("../styled-components");
styled_components_1.injectGlobal(templateObject_1 || (templateObject_1 = __makeTemplateObject(["", ""], ["", ""])), perfect_scrollbar_css_1.default && perfect_scrollbar_css_1.default.toString());
var StyledScrollWrapper = styled_components_1.default.div(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n  position: relative;\n"], ["\n  position: relative;\n"])));
var PerfectScrollbar = /** @class */ (function (_super) {
    __extends(PerfectScrollbar, _super);
    function PerfectScrollbar() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.handleRef = function (ref) {
            _this._container = ref;
        };
        return _this;
    }
    PerfectScrollbar.prototype.componentDidMount = function () {
        this.inst = new perfect_scrollbar_1.default(this._container, this.props.options || {});
    };
    PerfectScrollbar.prototype.componentDidUpdate = function () {
        this.inst.update();
    };
    PerfectScrollbar.prototype.componentWillUnmount = function () {
        this.inst.destroy();
    };
    PerfectScrollbar.prototype.render = function () {
        var _a = this.props, children = _a.children, className = _a.className, updateFn = _a.updateFn;
        updateFn(this.componentDidUpdate.bind(this));
        return (React.createElement(StyledScrollWrapper, { className: "scrollbar-container " + className, innerRef: this.handleRef }, children));
    };
    return PerfectScrollbar;
}(React.Component));
exports.PerfectScrollbar = PerfectScrollbar;
var templateObject_1, templateObject_2;
