// redoc/lib/components/Parameters/Parameters.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var DropdownOrLabel_1 = require("../DropdownOrLabel/DropdownOrLabel");
var ParametersGroup_1 = require("./ParametersGroup");
var common_elements_1 = require("../../common-elements");
var MediaTypesSwitch_1 = require("../MediaTypeSwitch/MediaTypesSwitch");
var Schema_1 = require("../Schema");
function safePush(obj, prop, item) {
    if (!obj[prop]) {
        obj[prop] = [];
    }
    obj[prop].push(item);
}
var PARAM_PLACES = ['path', 'query', 'cookie', 'header'];
var Parameters = /** @class */ (function (_super) {
    __extends(Parameters, _super);
    function Parameters() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Parameters.prototype.orderParams = function (params) {
        var res = {};
        params.forEach(function (param) {
            safePush(res, param.in, param);
        });
        return res;
    };
    Parameters.prototype.render = function () {
        var _a = this.props, body = _a.body, _b = _a.parameters, parameters = _b === void 0 ? [] : _b;
        if (body === undefined && parameters === undefined) {
            return null;
        }
        var paramsMap = this.orderParams(parameters);
        var paramsPlaces = parameters.length > 0 ? PARAM_PLACES : [];
        var bodyContent = body && body.content;
        return (React.createElement("div", null,
            paramsPlaces.map(function (place) { return (React.createElement(ParametersGroup_1.ParametersGroup, { key: place, place: place, parameters: paramsMap[place] })); }),
            bodyContent && React.createElement(BodyContent, { content: bodyContent })));
    };
    return Parameters;
}(React.PureComponent));
exports.Parameters = Parameters;
function DropdownWithinHeader(props) {
    return (React.createElement(common_elements_1.UnderlinedHeader, { key: "header" },
        "Request Body schema: ",
        React.createElement(DropdownOrLabel_1.DropdownOrLabel, __assign({}, props))));
}
function BodyContent(props) {
    var content = props.content;
    return (React.createElement(MediaTypesSwitch_1.MediaTypesSwitch, { content: content, renderDropdown: DropdownWithinHeader }, function (_a) {
        var schema = _a.schema;
        return React.createElement(Schema_1.Schema, { skipReadOnly: true, key: "schema", schema: schema });
    }));
}
